crates/wasi/src/preview2/poll.rs

```diff
@@ -0,0 +1,156 @@
+pub type ClosureFuture = Box<dyn Fn() -> PollableFuture<'static> + Send + Sync + 'static>;
+
+pub enum HostPollable {
+    TableEntry { index: u32, make_future: MakeFuture },
+    Closure(ClosureFuture),
+}
+
+pub trait TablePollableExt {
+    fn push_host_pollable(&mut self, p: HostPollable) -> Result<u32, TableError>;
+    fn get_host_pollable_mut(&mut self, fd: u32) -> Result<&mut HostPollable, TableError>;
+    fn delete_host_pollable(&mut self, fd: u32) -> Result<HostPollable, TableError>;
+}
+
+impl TablePollableExt for Table {
+    fn push_host_pollable(&mut self, p: HostPollable) -> Result<u32, TableError> {
+        self.push(Box::new(p))
+    }
+    fn get_host_pollable_mut(&mut self, fd: u32) -> Result<&mut HostPollable, TableError> {
+        self.get_mut::<HostPollable>(fd)
+    }
+    fn delete_host_pollable(&mut self, fd: u32) -> Result<HostPollable, TableError> {
+        self.delete::<HostPollable>(fd)
+    }
+}
+
+#[async_trait::async_trait]
+impl<T: WasiView> poll::Host for T {
+    async fn drop_pollable(&mut self, pollable: Pollable) -> Result<()> {
+        self.table_mut().delete_host_pollable(pollable)?;
+        Ok(())
+    }
+
+    async fn poll_oneoff(&mut self, pollables: Vec<Pollable>) -> Result<Vec<bool>> {
+        type ReadylistIndex = usize;
+
+        let table = self.table_mut();
+
+        let mut table_futures: HashMap<u32, (MakeFuture, Vec<ReadylistIndex>)> = HashMap::new();
+        let mut closure_futures: Vec<(PollableFuture<'_>, Vec<ReadylistIndex>)> = Vec::new();
+
+        for (ix, p) in pollables.iter().enumerate() {
+            match table.get_host_pollable_mut(*p)? {
+                HostPollable::Closure(f) => closure_futures.push((f(), vec![ix])),
+                HostPollable::TableEntry { index, make_future } => {
+                    match table_futures.entry(*index) {
+                        Entry::Vacant(v) => {
+                            v.insert((*make_future, vec![ix]));
+                        }
+                        Entry::Occupied(mut o) => {
+                            let (_, v) = o.get_mut();
+                            v.push(ix);
+                        }
+                    }
+                }
+            }
+        }
+
+        for (entry, (make_future, readylist_indices)) in table.iter_entries(table_futures) {
+            let entry = entry?;
+            closure_futures.push((make_future(entry), readylist_indices));
+        }
+
+        struct PollOneoff<'a> {
+            elems: Vec<(PollableFuture<'a>, Vec<ReadylistIndex>)>,
+        }
+        impl<'a> Future for PollOneoff<'a> {
+            type Output = Result<Vec<bool>>;
+
+            fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+                let mut any_ready = false;
+                let mut results = vec![false; self.elems.len()];
+                for (fut, readylist_indicies) in self.elems.iter_mut() {
```

Comment:
I should mention that we wouldn't get the full benefit of `FuturesUnordered` here given the one-off nature of `poll-oneoff`, so it might not be worth it until/unless we replace `poll-oneoff` with a resource that's intended to be polled repeatedly.

Reply:
I definitely don't disagree about the benefits of `FuturesUnordered`, but I would personally say we should start off simpler here and see how things go. My guess is that a large number of futures would have many other inefficiencies other than the polling behavior here, and in the meantime, at least to me personally, I feel that a manual "poll the list" is easier to understand than the `FuturesUnordered` abstraction